### PR TITLE
 The webhook secrets are automated, changes made in the pkg/pipelines folder- GITOPS-188

### PR DESCRIPTION
**Pull request by ishitasequeira**
**What type of PR is this?**

> /kind design

**What does does this PR do / why we need it**:
It should create a random secret if user does not want to create one on its own.

**Which issue(s) this PR fixes**:
https://issues.redhat.com/browse/GITOPS-188

Co-authored by  @gaganhegde and @ishitasequeira

### CODE REVIEW

File: pkg/odo/cli/pipelines/init.go [hunk 1]
```diff
@@ -108,7 +115,6 @@ func NewCmdInit(name, fullName string) *cobra.Command {
 	initCmd.Flags().StringVar(&o.gitOpsRepoURL, "gitops-repo-url", "", "GitOps repository e.g. https://github.com/organisation/repository")
 	initCmd.MarkFlagRequired("gitops-repo-url")
 	initCmd.Flags().StringVar(&o.gitOpsWebhookSecret, "gitops-webhook-secret", "", "provide the GitHub webhook secret for GitOps repository")
```
Comment (bigkevmcd) [inline pkg/odo/cli/pipelines/init.go]: Maybe for these, we should mention that we'll automatically generate the secret if it's not provided?

File: pkg/odo/cli/pipelines/bootstrap.go [hunk 2]
```diff
@@ -121,3 +135,17 @@ func removeEmptyStrings(s []string) []string {
 	}
 	return nonempty
 }
+
+//GenerateSecureString creates a webhook secret for you.
```
Comment (bigkevmcd) [inline pkg/odo/cli/pipelines/bootstrap.go]: Is that really what `GenerateSecureString` is doing?

Also, should it be exported?  I tend to fall on the side of not exporting things until I find a need to export.

Nitpick: I think this should have a space between `//` and `Generate`

File: pkg/odo/cli/pipelines/init.go [hunk 3]
```diff
@@ -69,6 +69,13 @@ func (io *InitParameters) Validate() error {
 	if len(removeEmptyStrings(strings.Split(gr.Path, "/"))) != 2 {
 		return fmt.Errorf("repo must be org/repo: %s", strings.Trim(gr.Path, ".git"))
 	}
+	if io.gitOpsWebhookSecret == "" {
+		gitSecret, err := GenerateSecureString()
+		if err != nil {
+			return err
```
Comment (bigkevmcd) [inline pkg/odo/cli/pipelines/init.go]: This error could be improved...the error you'll get back from `GenerateSecureString` will likely be a technical error, which may not be suitable for users?

Imagine if you ran the tool and it failed with an error that said something like `crypto/rand: blocked for 60 seconds waiting to read random data from the kernel`...huh?!

**Review by bigkevmcd — CHANGES_REQUESTED** [review]
I think it's also missing the same functionality for `add-service`.

You might run into trouble with `GenerateSecureString` at this point, and it will probably require exporting (and probably moving too).

File: pkg/odo/cli/pipelines/bootstrap.go [hunk 4]
```diff
@@ -63,6 +65,20 @@ func (io *BootstrapParameters) Validate() error {
 	if len(removeEmptyStrings(strings.Split(gr.Path, "/"))) != 2 {
 		return fmt.Errorf("repo must be org/repo: %s", strings.Trim(gr.Path, ".git"))
 	}
+	if io.GitOpsWebhookSecret == "" {
+		gitopsSecret, err := GenerateSecureString()
```
Comment (wtam2018) [inline pkg/odo/cli/pipelines/bootstrap.go]: gitOpsSecret as variable name
generateSecureString as func name (i.e. I don't think it nneds to be exported.)

File: pkg/odo/cli/pipelines/bootstrap.go [hunk 5]
```diff
@@ -121,3 +135,17 @@ func removeEmptyStrings(s []string) []string {
 	}
 	return nonempty
 }
+
+//GenerateSecureString creates a webhook secret for you.
+func GenerateSecureString() (string, error) {
```
Comment (wtam2018) [inline pkg/odo/cli/pipelines/bootstrap.go]: You should add some tests for this baby.

Comment (ishitasequeira) [reply pkg/odo/cli/pipelines/bootstrap.go:140]: I will make that change

File: pkg/odo/cli/pipelines/bootstrap.go (hunk 4 above)
Comment (ishitasequeira) [reply pkg/odo/cli/pipelines/bootstrap.go]: We would need to export it as we would need to have a check in service.go

Comment (bigkevmcd) [reply pkg/odo/cli/pipelines/bootstrap.go]: We've discussed this, and I think we can move it to `pkg/utils/secrets.go` with `pkg/utils/secrets_test.go`

The test should generate a fixed number of items, add them to `map[string]bool` and at the end, check that the `len()` of that map is the same as the fixed number.

Oh, and the test should also iterate over all the strings returned, and check that they're all len() == 20.

File: pkg/odo/cli/pipelines/init.go (hunk 1 above)
Comment (gaganhegde) [reply pkg/odo/cli/pipelines/init.go]: Made the changes.

File: pkg/odo/cli/pipelines/bootstrap.go (hunk 5 above)
Comment (gaganhegde) [reply pkg/odo/cli/pipelines/bootstrap.go]: Haha..!!Test case added.

File: pkg/odo/cli/pipelines/bootstrap.go (hunk 4 above)
Comment (gaganhegde) [reply pkg/odo/cli/pipelines/bootstrap.go]: Test case with the requirements mentioned above completed.

File: pkg/odo/cli/pipelines/bootstrap.go (hunk 2 above)
Comment (gaganhegde) [reply pkg/odo/cli/pipelines/bootstrap.go]: Comment modified to match the purpose of the function.

File: pkg/util/generate_secret_test.go [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+		}
+		if err == nil {
+			uniqueSecrets[secretKey] = true
+		}
+	}
+	want := 5
+	got := len(uniqueSecrets)
+
+	if want != got {
+		t.Error("The secrets generated by GenerateSecureString() arent unique")
```
Comment (bigkevmcd) [inline pkg/util/generate_secret_test.go]:
```suggestion
		t.Errorf("The secrets generated by GenerateSecureString() arent unique, got %d unique secrets", len(uniqueSecrets))
```

Comment (bigkevmcd) [reply pkg/util/generate_secret_test.go]: But actually....

if l := len(uniqueSecrets); l != 5 {
    t.Errorf("The secrets generated by GenerateSecureString() arent unique, got %d unique secrets", l)
}

Comment (ishitasequeira) [reply pkg/util/generate_secret_test.go]: made the above change

File: pkg/odo/cli/pipelines/service/add.go [hunk 7]
```diff
@@ -85,7 +85,7 @@ func newCmdAdd(name, fullName string) *cobra.Command {
 	}
 
 	cmd.Flags().StringVar(&o.gitRepoURL, "git-repo-url", "", "source Git repository URL")
-	cmd.Flags().StringVar(&o.webhookSecret, "webhook-secret", "", "source Git repository webhook secret")
+	cmd.Flags().StringVar(&o.webhookSecret, "webhook-secret", "", "source Git repository webhook secret (if not provided, it will be auto-generated)")
```
Comment (bigkevmcd) [inline pkg/odo/cli/pipelines/service/add.go:88]: I think this needs more than just a comment.

**Review by bigkevmcd — APPROVED** [review]

**Comment by bigkevmcd** [comment]
I find the split between service/add and bootstrap's secret generation a bit hard to follow, but I understand why.

I think we should move the secret generation to the same layer (i.e. pkg/pipelines), but this solves the problem we have for now.

File: pkg/pipelines/service.go [hunk 8]
```diff
@@ -69,7 +69,12 @@ func serviceResources(m *config.Manifest, fs afero.Fs, p *AddServiceParameters)
...
 	cfg := m.GetPipelinesConfig()
 	if cfg != nil && p.WebhookSecret == "" && p.GitRepoURL != "" {
-		return nil, fmt.Errorf("The webhook secret is required")
+		gitSecret, err := util.GenerateSecureString(20)
+		if err != nil {
+			return nil, err
+		}
+		p.WebhookSecret = gitSecret
+
 	}
```
Comment (wtam2018) [inline pkg/pipelines/service.go:78]: Any reasons why `geenerate secret` is done here in the backend for `service add` command but `bootstrap` and `init` commands are in the cli?   In fact, this validation should probably have been done in the CLI's Validate().

Comment (gaganhegde) [reply pkg/pipelines/service.go:78]: Yes William, the config retrieved form the manifest exists at the pkg/pipelines level.
The end to end testing is now complete.
Kindly review and approve changes.

File: pkg/pipelines/bootstrap.go [hunk 9]
```diff
@@ -46,6 +46,20 @@ type BootstrapOptions struct {
 
 // Bootstrap bootstraps a GitOps pipelines and repository structure.
 func Bootstrap(o *BootstrapOptions, appFs afero.Fs) error {
+	if o.GitOpsWebhookSecret == "" {
+		gitopsSecret, err := secrets.GenerateSecureString(20)
```
Comment (wtam2018) [inline pkg/pipelines/bootstrap.go]: You probably should make a constant for the len (20).   Also, I would rename the function to secrets.GenerateString()

File: pkg/pipelines/secrets/generate_secret.go [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+)
+
+const charset = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789$:#"
+
+// GenerateSecureString creates a random secret of the desired length specifically for the webhook secrets.
+func GenerateSecureString(length int) (string, error) {
+	b := make([]byte, length)
+	_, err := rand.Read(b)
+	if err != nil {
+		return "", fmt.Errorf("Failed to generate secure webhook secret")
```
Comment (wtam2018) [inline pkg/pipelines/secrets/generate_secret.go]: I wouldn't mention "secure webhook" in the error because this method can serve other purposes than creating secure for webhook.

Comment (gaganhegde) [reply pkg/pipelines/secrets/generate_secret.go]: Changes made to the more generic error message.

File: pkg/pipelines/bootstrap.go (hunk 9 above)
Comment (gaganhegde) [reply pkg/pipelines/bootstrap.go]: Changes made to incorporate constant and name of function changed.

**Review by wtam2018 — APPROVED** [review]
